feat: Add Token Credential Support

elliot-huffman opened the pull request:
## Overview

Adds support for the base token credential authentication that is implemented by [`@Azure/Identity`](https://www.npmjs.com/package/@azure/identity).
This PR does not remove support for the other [`@Azure/Identity`](https://www.npmjs.com/package/@azure/identity) methods.

The other auth methods are now able to be removed as the credential chain type supports whatever the end user throws at it. Default, service principal, and any combination of crazy auth methods and all future ones that MSFT adds to the MSAL system.
All [`@Azure/Identity`](https://www.npmjs.com/package/@azure/identity) auth classes derive from the credential chain class so all the other existing authentication methods are supported through this specific option as well as 100% of their configurations with no additional code that is required from us.

See #1623 for more info.
This would be the core auth system work required to solve issues like #1144 as they would be able to use auth that makes sense for them.


## Sample Documentation

- `microsoft-credential-chain`
```TypeScript
const msAuth = new DefaultAzureCredential()

"authentication": {
  "type":'microsoft-credential-chain',
  ...,
  "options": {
    "credential": msAuth
  }
}
```
- credential (TokenCredential): Instance of a token credential from a package like [`@Azure/Identity`](https://www.npmjs.com/package/@azure/identity). This could be any of the built-in credential types such as `DefaultAzureCredential`, `EnvironmentCredential`, `ChainedTokenCredentia`, `VisualStudioCodeCredential`, etc. All the MS supported classes derive from the `TokenCredential` interface so all the pre-built auth methods are supported by inheritance. See the @Azure/Identity package's documentation (https://www.npmjs.com/package/@azure/identity) for more ways to customize your authentication experience for your environment's specifics.

(optional section)
**Best Practices**
Migrate from other `azure-active-directory` prefixed auth config types to this one as they will be removed in the future to simplify the code base.

arthurschreiber reviewed (commented):
Sweet, this looks great! Can you fix the formatting to use 2 spaces instead of tabs? 🙇

elliot-huffman commented:
> Sweet, this looks great! Can you fix the formatting to use 2 spaces instead of tabs? 🙇

Fixed!

arthurschreiber commented on src/connection.ts:
```diff
@@ -290,6 +291,17 @@ interface AzureActiveDirectoryServicePrincipalSecret {
 
+/** Structure that defines the options that are necessary to authenticate the tedious instance with an `@azure/identity` chained token credential derived class. */
+interface CredentialChainAuthentication {
+    /** Unique designator for the type of authentication to be used. */
```
I think this is still using tabs instead of spaces. 😬

elliot-huffman replied:
Sorry about that, I tried to regex search `/t` in GitHub but I found nothing. Should be fixed now.

elliot-huffman replied:
@arthurschreiber, I found one other place where the formatting might not be correct. Fixed that too.

elliot-huffman replied:
In the future, I will 100% not be free handing this stuff in the GitHub file editor. I will follow the PR instructions and do it locally :-P

arthurschreiber commented:
Would you mind updating the lock file as well?

elliot-huffman commented:
> Would you mind updating the lock file as well?

That would be important. One sec.

elliot-huffman commented:
I believe it is fixed now.
I found an unrelated bug when I added some types to variables as part of this change. It was a bug where the access token could be null, so I added a type guard that triggered the correct error if this case is ever reached.

arthurschreiber commented:
I had some more time to look at this and think about this. 😅 

How do you feel about renaming the credential method to `azure-identity-token-credential`, and change this to only use the `TokenCredential` interface instead? `TokenCredential` is the _actual_ thing we're interested in - we don't actually care whether it's a chained token credential or not (and apparently there's various token credentials that don't inherit from `ChaintedTokenCredential`: https://github.com/search?q=repo%3AAzure%2Fazure-sdk-for-js%20%22implements%20TokenCredential%22&type=code)

arthurschreiber commented:
Also, in general I'd be in favor of deprecating (and removing) the other azure identity related methods - it would allow us to completely drop the runtime dependency on `@azure/identity`, resolving some long standing issues around startup performance.

arthurschreiber commented on src/connection.ts:
```diff
@@ -290,6 +293,17 @@ interface AzureActiveDirectoryServicePrincipalSecret {
+  /** Set of configurations that are required or allowed with this authentication type. */
+  options: {
+    /** Credential object used to authenticate to the resource. */
+    credential: ChainedTokenCredential;
```
I think we should change the type here to `credential: TokenCredential | (): TokenCredential | (): Promise<TokenCredential>`

This would allow clients to re-use the connection configuration and re-generate tokens whenever a connection is opened with the same configuration automatically.

elliot-huffman replied:
Also, since `Azure AD` is the old name of Microsoft's directory service. I think `microsoft-token-credential` would be a better name since it would be future proofed for when MSFT eventually renames the or changes the terminology.

What do you think?

elliot-huffman replied:
> I think we should change the type here to `credential: TokenCredential | () => TokenCredential | () => Promise<TokenCredential>`
> 
> This would allow clients to re-use the connection configuration and re-generate tokens whenever a connection is opened with the same configuration automatically.

Also, did I read this right? do you want me to create some sort of function invoker thingy?

arthurschreiber replied:
Yeah, but we can also just skip this for now and add it later. Sorry for the confusion 🙇‍♂️

elliot-huffman replied:
I will make a second Issue for this for this behavior.

elliot-huffman commented:
> Also, in general I'd be in favor of deprecating (and removing) the other azure identity related authentication methods - it would allow us to completely drop the runtime dependency on `@azure/identity`, resolving some long standing issues around startup performance.

I believe the actual constructor validation that I wrote is for token credential instead of change credential, as I was testing with some non-chained credentials it failed and I had to update the validation to support it properly.
The documentation I wrote was definitely for change credentials but I will update that wording even though the back end is looking for the token credential structure. 
The token credential class is not an actual class, it's just an interface, the constructor type guard I wrote is for validating that interface. I found out the hard way when I tried to do an instanceof keyword against it, LOL.

elliot-huffman commented:
> I had some more time to look at this and think about this. 😅 
> 
> How do you feel about renaming the credential method to `azure-identity-token-credential`, and change this to only use the `TokenCredential` interface instead?
> 
> `TokenCredential` is the _actual_ thing we're interested in - we don't actually care whether it's a chained token credential or not (and apparently there's various token credentials that don't inherit from `ChainedTokenCredential`: https://github.com/search?q=repo%3AAzure%2Fazure-sdk-for-js%20%22implements%20TokenCredential%22&type=code)

I can definitely rename it :-)

elliot-huffman commented:
Just read through the lint report, this one is not my fault:
![image](https://github.com/tediousjs/tedious/assets/5030490/c786323e-4a34-4fef-8823-d29c8a289a04)

I am planning to overhaul the linting system with a future PR and this would be fixed at that time.

If you are ok with it, would you be able to override this one?

arthurschreiber commented on src/connection.ts:
```diff
@@ -3094,10 +3120,7 @@ class Connection extends EventEmitter {
-   * @private
-   */
-  makeRequest(request: Request | BulkLoad, packetType: number, payload: (Iterable<Buffer> | AsyncIterable<Buffer>) & { toString: (indent?: string) => string }) {
+  private makeRequest(request: Request | BulkLoad, packetType: number, payload: (Iterable<Buffer> | AsyncIterable<Buffer>) & { toString: (indent?: string) => string }) {
```
Do you mind switching this back to not be `private`? This is breaking some unrelated code.

elliot-huffman replied:
interesting, it fails a compile if I keep it as a JS Doc though, specifically 'type annotaiton not expected'

elliot-huffman commented on src/connection.ts:
```diff
@@ -290,6 +291,17 @@ interface AzureActiveDirectoryServicePrincipalSecret {
+/** Structure that defines the options that are necessary to authenticate the Tedious.JS instance with an `@azure/identity` token credential. */
+interface TokenCredentialAuthentication {
+  /** Unique designator for the type of authentication to be used. */
+  type: 'microsoft-token-credential';
```
I just want to keep the namespace from anything else that could be a token credential in the future that isn't from MSFT but I do also see that this is a low possibility, I personally sorta lean towards keeping it prefixed with `microsoft` but I will remove it if you still lean the other way. I am not that strongly opiniated on it too.

elliot-huffman commented:
It looks like it ran successfully, it is just failing to connect to codecov? I could be wrong; I don't use that service yet. @arthurschreiber, any todo items for me?

arthurschreiber commented:
I'm trying to figure out what's going on with codecov, seems like it needs to be updated to the latest version of the action (see https://github.com/tediousjs/tedious/pull/1627).

arthurschreiber commented:
Would you be okay with me taking the changes done so far in here and opening a new PR? I'd like to make sure we have tests running for this new functionality, but azure tests don't run on pull requests originating from forks because of security restrictions.

elliot-huffman commented:
> Would you be okay with me taking the changes done so far in here and opening a new PR? I'd like to make sure we have tests running for this new functionality, but azure tests don't run on pull requests originating from forks because of security restrictions.

On the condition that I get recognition, my ego demands it :-P

maybe being added to the contributors section of the `package.json` as "Elliot Huffman <ehuffman@elliot-labs.com> (https://elliot-labs.com)" or something like that.

arthurschreiber commented:
The commit(s) will stay attributed to you. 😬

elliot-huffman commented:
> The commit(s) will stay attributed to you. 😬

my bad, works for me!
